ddtrace/settings/_inferred_base_service.py

```diff
@@ -0,0 +1,203 @@
+import fnmatch
+import os
+import pathlib
+import re
+import sys
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+
+INIT_PY = "__init__.py"
+ALL_PY_FILES = "*.py"
+
+CACHE: Dict[str, Optional[str]] = {}
+
+
+class DetectionContext:
+    def __init__(self, envs):
+        self.envs = envs
+
+
+class ServiceMetadata:
+    def __init__(self, name):
+        # prevent service name from being an empty string
+        self.name = name if name != "" else None
+
+
+class PythonDetector:
+    def __init__(self, ctx):
+        self.ctx = ctx
+        name = "python"
+        self.name = name
+
+        # This pattern matches:
+        # - Starts with an optional directory (anything before the last '/' or '')
+        # - Ends with the expected command name, possibly followed by a version
+        # - Ensures that it does not end with .py
+        # - Match /python, /python3.7, etc.
+        self.pattern = r"(^|/)(?!.*\.py$)(" + re.escape(name) + r"(\d+\.\d+)?$)"
+
+    def detect(self, args: List[str]) -> Tuple[ServiceMetadata, bool]:
+        """
+        Detects and returns service metadata based on the provided list of arguments.
+
+        This function iterates through the provided arguments, skipping any that are
+        flags (starting with '-') or environment variable assignments (containing '=').
+        If a valid module flag ('-m') is encountered, it will switch to module detection mode.
+        It checks for existing directories and deduces package names from provided paths
+        to generate service metadata.
+
+        Args:
+            args (List[str]): A list of command-line arguments.
+
+        Returns:
+            Tuple[ServiceMetadata, bool]: A tuple containing:
+                - ServiceMetadata: The detected service metadata.
+                - bool: A flag indicating whether valid metadata was found.
+        """
+        prev_arg_is_flag = False
+        module_flag = False
+
+        for arg in args:
+            has_flag_prefix = arg.startswith("-")
+            is_env_variable = "=" in arg
+
+            should_skip_arg = prev_arg_is_flag or has_flag_prefix or is_env_variable
+
+            if module_flag:
+                return ServiceMetadata(arg), True
+
+            if not should_skip_arg:
+                abs_path = pathlib.Path(arg).resolve()
+                if not abs_path.exists():
+                    continue
+                stripped = abs_path
+                if not stripped.is_dir():
+                    stripped = stripped.parent
+                value, ok = self.deduce_package_name(stripped)
+                if ok:
+                    return ServiceMetadata(value), True
+                return ServiceMetadata(self.find_nearest_top_level(stripped)), True
+
+            if has_flag_prefix and arg == "-m":
+                module_flag = True
+
+            prev_arg_is_flag = has_flag_prefix
+
+        return ServiceMetadata(""), False
```

Comment:
should we just return `Optional[ServiceMetadata]` instead? and then this is just `return None`

Reply:
smart, done.